kernel/src/mm/guestmem.rs

```diff
@@ -271,6 +277,98 @@ impl<T: Copy> InsnMachineMem for GuestPtr<T> {
     }
 }
 
+#[derive(Debug)]
+pub struct UserPtr<T: Copy + FromBytes> {
+    guest_ptr: GuestPtr<T>,
+}
+
+impl<T: Copy + FromBytes> UserPtr<T> {
+    #[inline]
+    pub fn new(v: VirtAddr) -> Self {
+        Self {
+            guest_ptr: GuestPtr::new(v),
+        }
+    }
+
+    fn check_bounds(&self) -> bool {
+        let v = VirtAddr::from(self.guest_ptr.ptr);
+        if !(USER_MEM_START..USER_MEM_END).contains(&v)
+            || !(USER_MEM_START..USER_MEM_END).contains(&(v + size_of::<T>()))
+        {
+            return false;
+        }
+        true
+    }
+
+    /// # Safety
+    ///
+    /// This function ensures that the address is valid for reading. If the address is not valid,
+    /// it returns an `Err(SvsmError::InvalidAddress)`.
+    #[inline]
+    pub unsafe fn read(&self) -> Result<T, SvsmError> {
```

Comment:
```suggestion
    #[inline]
    pub fn read(&self) -> Result<T, SvsmError> {
```

This function (and `read_c_string`, `write`, and `write_ref`) don't need to be unsafe. We have all the checks in place to make sure that nothing bad can happen :)